src/core/utils/Task.cpp

```diff
@@ -34,42 +34,57 @@ std::shared_ptr<Task> Task::Run(std::function<void()> block) {
 Task::Task(std::function<void()> block) : block(std::move(block)) {
 }
 
+bool Task::waiting() {
+  return status.load(std::memory_order_relaxed) == TaskStatus::waiting;
+}
+
 bool Task::executing() {
-  std::lock_guard<std::mutex> autoLock(locker);
-  return _executing;
+  return status.load(std::memory_order_relaxed) == TaskStatus::executing;
 }
 
 bool Task::cancelled() {
-  std::lock_guard<std::mutex> autoLock(locker);
-  return _cancelled;
+  return status.load(std::memory_order_relaxed) == TaskStatus::cancelled;
 }
 
 bool Task::finished() {
-  std::lock_guard<std::mutex> autoLock(locker);
-  return !_executing && !_cancelled;
+  return status.load(std::memory_order_relaxed) == TaskStatus::finished;
 }
 
 void Task::wait() {
-  std::unique_lock<std::mutex> autoLock(locker);
-  if (!_executing) {
+  auto oldStatus = status.load(std::memory_order_relaxed);
+  if (oldStatus == TaskStatus::cancelled || oldStatus == TaskStatus::finished) {
     return;
   }
-  condition.wait(autoLock);
+  if (oldStatus == TaskStatus::waiting) {
+    if (status.compare_exchange_weak(oldStatus, TaskStatus::executing, std::memory_order_acq_rel,
+                                     std::memory_order_relaxed)) {
+      block();
```

Comment:
这句之前加一行注释：If wait() is called from the thread pool, all threads might block, leaving no thread to execute this task. To avoid deadlock, execute the task directly on the current thread if it's queued.